File: app/src/main/java/com/tasomaniac/openwith/preferred/AppRemoveDialogFragment.java
package com.tasomaniac.openwith.preferred;

import android.app.Dialog;
import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialogFragment;
import android.text.Html;

import com.tasomaniac.openwith.R;
import com.tasomaniac.openwith.resolver.DisplayActivityInfo;

public class AppRemoveDialogFragment extends AppCompatDialogFragment {

    private static final String EXTRA_INFO = "EXTRA_INFO";

    static AppRemoveDialogFragment newInstance(DisplayActivityInfo info) {
        AppRemoveDialogFragment fragment = new AppRemoveDialogFragment();
        Bundle args = new Bundle();
        args.putParcelable(EXTRA_INFO, info);
        fragment.setArguments(args);
        return fragment;
    }

    private Callbacks callbacks;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        callbacks = (Callbacks) context;
    }

    @Override
    public void onDetach() {
        callbacks = null;
        super.onDetach();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        DisplayActivityInfo info = getInfo();
        CharSequence message = appRemoveDialogContentFrom(getResources(), info);

        return new AlertDialog.Builder(getActivity())
                .setTitle(R.string.title_remove_preferred)
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> callbacks.onAppRemoved(info))
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    private DisplayActivityInfo getInfo() {
        final DisplayActivityInfo info = getArguments().getParcelable(EXTRA_INFO);
        if (info == null) {
            throw new IllegalArgumentException("Use newInstance method to create the Dialog");
        }
        return info;
    }

    private static CharSequence appRemoveDialogContentFrom(Resources resources, DisplayActivityInfo info) {
        String content = resources.getString(
                R.string.message_remove_preferred,
                info.displayLabel(),
                info.extendedInfo(),
                info.extendedInfo()
        );
        //noinspection deprecation
        return Html.fromHtml(content);
    }

    interface Callbacks {
        void onAppRemoved(DisplayActivityInfo info);
    }
}

File: app/src/main/java/com/tasomaniac/openwith/preferred/AppRemoveDialogFragment.kt
package com.tasomaniac.openwith.preferred

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.core.os.bundleOf
import androidx.core.text.parseAsHtml
import com.tasomaniac.openwith.R
import com.tasomaniac.openwith.resolver.DisplayActivityInfo

class AppRemoveDialogFragment : AppCompatDialogFragment() {

    private var callbacks: Callbacks? = null

    private val info: DisplayActivityInfo
        get() = arguments!!.getParcelable(EXTRA_INFO)!!

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callbacks = context as Callbacks
    }

    override fun onDetach() {
        callbacks = null
        super.onDetach()
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val message = resources.getString(
            R.string.message_remove_preferred,
            info.displayLabel, info.extendedInfo, info.extendedInfo
        ).parseAsHtml()

        return AlertDialog.Builder(activity!!)
            .setTitle(R.string.title_remove_preferred)
            .setMessage(message)
            .setPositiveButton(android.R.string.ok) { _, _ -> callbacks!!.onAppRemoved(info) }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    interface Callbacks {
        fun onAppRemoved(info: DisplayActivityInfo)
    }

    companion object {

        private const val EXTRA_INFO = "EXTRA_INFO"

        internal fun newInstance(info: DisplayActivityInfo) =
            AppRemoveDialogFragment().apply {
                arguments = bundleOf(EXTRA_INFO to info)
            }
    }
}
